libfastpm/gravity.c

```diff
@@ -413,13 +414,40 @@ _fastpm_solver_compute_force(FastPMSolver * fastpm,
 
 }
 
+static double
+lra_neutrinos(double k, nu_lra_power * nulra)
+{
+    /*Add neutrino power if desired*/
+    if(k <= 0)
+        return 1;
+    /* Change the units of k to match those of logkk*/
+    double logk = log(k);
+    /* Floating point roundoff and the binning means there may be a mode just beyond the box size.*/
+    if(logk < nulra->logknu[0] && logk > nulra->logknu[0]-log(2) )
+        logk = nulra->logknu[0];
+    else if( logk > nulra->logknu[nulra->size-1])
+        logk = nulra->logknu[nulra->size-1];
+    /* Note get_neutrino_powerspec returns Omega_nu / (Omega0 -OmegaNu) * delta_nu / P_cdm^1/2, which is dimensionless.
+        * So below is: M_cdm * delta_cdm (1 + Omega_nu/(Omega0-OmegaNu) (delta_nu / delta_cdm))
+        *            = M_cdm * (delta_cdm (Omega0 - OmegaNu)/Omega0 + Omega_nu/Omega0 delta_nu) * Omega0 / (Omega0-OmegaNu)
+        *            = M_cdm * Omega0 / (Omega0-OmegaNu) * (delta_cdm (1 - f_nu)  + f_nu delta_nu) )
+        *            = M_cdm * Omega0 / (Omega0-OmegaNu) * delta_t
+        *            = (M_cdm + M_nu) * delta_t
+        * This is correct for the forces, and gives the right power spectrum,
+        * once we multiply PowerSpectrum.Norm by (Omega0 / (Omega0 - OmegaNu))**2 */
+    double delta_nu = gsl_interp_eval(nulra->nu_spline, nulra->logknu, nulra->delta_nu_ratio, logk, NULL);
+    const double nufac = 1 + nulra->nu_prefac * delta_nu;
+    return nufac;
```

Comment:
I'm not sure I follow the comment on line 430. What is `get_neutrino_powerspec`? Do you mean `nulra->nu_prefac * delta_nu$` on line 438? Is the derivation in the comment what the function outputs?

By my calculation, in line 439, `delta_nu` = $\delta_\nu/P_c^{0.5}$, `nulra->nu_prefac` = $\Omega_\nu(a)/\Omega_c(a)$, so this function returns $1 + \frac{\Omega_\nu(a)}{\Omega_c(a)}\frac{\delta_\nu}{P_c^{0.5}} = 1 + \frac{f_\nu\delta_\nu}{f_c\delta_c} = \frac{f_c\delta_c+f_\nu\delta_\nu}{f_c\delta_c} = \frac{\delta_m}{f_c\delta_c}$.

Then in line 505 of gravity.c this is applied to transfer $\delta_c$. This gives $\delta_c \cdot \frac{\delta_m}{f_c\delta_c} = \frac{\delta_m}{f_c}$.

Is this correct and what we want, or do we want $\delta_m$? Perhaps this extra $f_c$ factor is related to why there's a factor of `MtotbyMcdm` in `powerpectrum_save_nu`?

Reply:
Yes, your calculation of $\delta_\nu$ is correct. In MP-Gadget the transfer is applied (in the equivalent of gravity.c) to $M_c * \delta_c$ (like the comment says). Then we need to divide by the total mass to get the output. If fastpm is actually computing $\delta_c$ rather than $M_c delta_c$ then indeed we should change the prefactors.